src/compiler/checker.ts

```diff
@@ -32422,29 +32422,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (!(type.flags & TypeFlags.Union)) {
             return getContextualCallSignature(type, node);
         }
-        let signatureList: Signature[] | undefined;
-        const types = (type as UnionType).types;
-        for (const current of types) {
-            const signature = getContextualCallSignature(current, node);
-            if (signature) {
-                if (!signatureList) {
-                    // This signature will contribute to contextual union signature
-                    signatureList = [signature];
-                }
-                else if (!compareSignaturesIdentical(signatureList[0], signature, /*partialMatch*/ false, /*ignoreThisTypes*/ true, /*ignoreReturnTypes*/ true, compareTypesIdentical)) {
-                    // Signatures aren't identical, do not use
-                    return undefined;
-                }
-                else {
-                    // Use this signature for contextual union signature
-                    signatureList.push(signature);
-                }
+
+        let signatures = mapDefined((type as UnionType).types, t => getContextualCallSignature(t, node));
+        if (hasSyntacticModifier(node, ModifierFlags.Async)) {
+            signatures = signatures.filter(s => !!getAwaitedTypeOfPromise(getReturnTypeOfSignature(s)));

```

Comment:
this might not be correct for async generators since they have `ModifierFlags.Async` but are not expected to return a type assignable to `Promise<any>`